crates/iceberg/src/spec/values.rs

```diff
@@ -70,6 +79,571 @@ pub enum PrimitiveLiteral {
+    ///
+    /// Example:
+    /// ```rust
+    /// use ordered_float::OrderedFloat;
+    /// use iceberg::spec::{Literal, PrimitiveLiteral, Datum};
+    /// let t = Datum::float( 32.1f32 );
+    ///
+    /// assert_eq!(&format!("{t}"), "32.1");
+    /// assert_eq!(Literal::Primitive(PrimitiveLiteral::Float(OrderedFloat(32.1))), t.into());
+    /// ```
+    pub fn float<T: Into<f32>>(t: T) -> Self {
+        Self {
+            r#type: PrimitiveType::Float,
+            literal: PrimitiveLiteral::Float(OrderedFloat(t.into())),
+        }
+    }
+
+    /// Creates an 32bit floating point number.
+    ///
+    /// Example:
+    /// ```rust
+    /// use ordered_float::OrderedFloat;
+    /// use iceberg::spec::{Literal, PrimitiveLiteral, Datum};
+    /// let t = Datum::double( 32.1f64 );
+    ///
+    /// assert_eq!(&format!("{t}"), "32.1");
+    /// assert_eq!(Literal::Primitive(PrimitiveLiteral::Double(OrderedFloat(32.1))), t.into());
+    /// ```
+    pub fn double<T: Into<f64>>(t: T) -> Self {
+        Self {
+            r#type: PrimitiveType::Double,
+            literal: PrimitiveLiteral::Double(OrderedFloat(t.into())),
+        }
+    }
+
+    /// Creates date literal from number of days from unix epoch directly.
+    ///
+    /// Example:
+    /// ```rust
+    ///
+    /// use iceberg::spec::{Literal, PrimitiveLiteral, Datum};
+    /// // 2 days after 1970-01-01
+    /// let t = Datum::date(2);
+    ///
+    /// assert_eq!(&format!("{t}"), "1970-01-03");
+    /// assert_eq!(Literal::Primitive(PrimitiveLiteral::Date(2)), t.into());
+    /// ```
+    pub fn date(days: i32) -> Self {
+        Self {
+            r#type: PrimitiveType::Date,
+            literal: PrimitiveLiteral::Date(days),
+        }
+    }
+
+    /// Creates a date in `%Y-%m-%d` format, assume in utc timezone.
+    ///
+    /// See [`NaiveDate::from_str`].
+    ///
+    /// Example
+    /// ```rust
+    /// use iceberg::spec::{Literal, Datum};
+    /// let t = Datum::date_from_str("1970-01-05").unwrap();
+    ///
+    /// assert_eq!(&format!("{t}"), "1970-01-05");
+    /// assert_eq!(Literal::date(4), t.into());
+    /// ```
+    pub fn date_from_str<S: AsRef<str>>(s: S) -> Result<Self> {
+        let t = s.as_ref().parse::<NaiveDate>().map_err(|e| {
+            Error::new(
+                ErrorKind::DataInvalid,
+                format!("Can't parse date from string: {}", s.as_ref()),
+            )
+            .with_source(e)
+        })?;
+
+        Ok(Self::date(date_from_naive_date(t)))
+    }
+
+    /// Create a date from calendar date (year, month and day).
+    ///
+    /// See [`NaiveDate::from_ymd_opt`].
+    ///
+    /// Example:
+    ///
+    ///```rust
+    /// use iceberg::spec::{Literal, Datum};
+    /// let t = Datum::date_from_ymd(1970, 1, 5).unwrap();
+    ///
+    /// assert_eq!(&format!("{t}"), "1970-01-05");
+    /// assert_eq!(Literal::date(4), t.into());
+    /// ```
+    pub fn date_from_ymd(year: i32, month: u32, day: u32) -> Result<Self> {
```

Comment:
![til](https://github.com/apache/iceberg-rust/assets/1134248/f2f3e16f-a168-48e9-ad0e-3a1162b8d655)

Rust does not have overloading 👍